internal/pkg/agent/storage/store/internal/migrations/migrations.go

```diff
@@ -0,0 +1,97 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package migrations
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"gopkg.in/yaml.v2"
+)
+
+type StateStore struct {
+	Action      Action        `yaml:"action"`
+	AckToken    string        `yaml:"ack_token"`
+	ActionQueue []ActionQueue `yaml:"action_queue"`
+}
+
+// Action is a struct to read an Action Policy Change from its old
+// YAML format.
+type Action struct {
+	ActionID     string                 `yaml:"action_id"`
+	Type         string                 `yaml:"action_type"`
+	StartTime    time.Time              `yaml:"start_time,omitempty"`
+	SourceURI    string                 `yaml:"source_uri,omitempty"`
+	RetryAttempt int                    `yaml:"retry_attempt,omitempty"`
+	Policy       map[string]interface{} `yaml:"policy,omitempty"`
+	IsDetected   bool                   `yaml:"is_detected,omitempty"`
+}
+
+// ActionQueue is a struct to read the action queue from its old YAML format.
+type ActionQueue struct {
+	ActionID       string                 `yaml:"action_id"`
+	Type           string                 `yaml:"type"`
+	StartTime      time.Time              `yaml:"start_time,omitempty"`
+	ExpirationTime time.Time              `yaml:"expiration,omitempty"`
+	Version        string                 `yaml:"version,omitempty"`
+	SourceURI      string                 `yaml:"source_uri,omitempty"`
+	RetryAttempt   int                    `yaml:"retry_attempt,omitempty"`
+	Policy         map[string]interface{} `yaml:"policy,omitempty"`
+	IsDetected     bool                   `yaml:"is_detected,omitempty"`
+}
+
+type loader interface {
+	Load() (io.ReadCloser, error)
+}
+
+// LoadActionStore loads an action store from loader.
+func LoadActionStore(loader loader) (*Action, error) {
+	return LoadStore[Action](loader)
+}
+
+// LoadYAMLStateStore loads the old YAML state store from loader.
+func LoadYAMLStateStore(loader loader) (*StateStore, error) {
+	return LoadStore[StateStore](loader)
+}
+
+// LoadStore loads a YAML file.
+func LoadStore[Store any](loader loader) (store *Store, err error) {
+	// Store is a generic type, this might be needed.
+	store = new(Store)
+
+	reader, err := loader.Load()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load action store: %w", err)
+	}
+	defer func() {
+		err2 := reader.Close()
+		if err2 != nil {
+			err2 = fmt.Errorf("migration storeLoad failed to close reader: %w", err2)
+		}
+		if err != nil {
+			err = errors.Join(err, err2)
+		}
```

Comment:
You can probably join the errors whether err is nil or not
```suggestion
		if err2 != nil {
			err = errors.Join(err, fmt.Errorf("migration storeLoad failed to close reader: %w", err2))
		}
```